### PR TITLE
add BS.2076-2 PositionOffset to audioObject

### LINKED ISSUE #89: Support positionOffset sub-element of audioObject
[2076-2](https://www.itu.int/dms_pubrec/itu-r/rec/bs/R-REC-BS.2076-2-201910-I!!PDF-E.pdf) adds a new sub-element positionOffset

It is optional and applied to each axis of the coordinate system in use. The axis is set via a coordinate attribute in the same way as the position subElement of audioBlockFormat.

> positionOffset sub-element of audioObject: This defines position offset values that should be applied to the position metadata of all audio corresponding to the parent audioObject. It describes the initial playback position offset of the audioObject during rendering. 

* What are valid ranges for these? For example Azimuth is restricted to +-180 for the Position sub-element. Should this be the same range? - Use same ranges as specified for block format positions. Distance offset should be able to be negative.

### PR DESCRIPTION
**Pull request by kristianhentschelbbc**
Closes #89.

- [x] Add PositionOffset classes
- [x] Can create and access spherical and cartesian position offsets on AudioObject
- [x] Can parse AudioObjects containing position offsets from XML
- [x] Can write AudioObjects containing position offsets to XML
- ~Include positionOffset in `AudioObject::print`?~
- [x] Make sure tests cover all new code
- ~Discuss if position offset and position should have a common ancestor to reduce code duplication~
- [x] Update change log

### CODE REVIEW

**Comment by tomjnixon** [comment]
This looks pretty good; a few comments:

- I'm not sure about the defaults and limits of the offsets; the specification doesn't give either. Defaulting to zero makes sense. If we're going to keep the limits then I'd say the elevation offset should be -180 to 180 (difference between sources at -90 and 90), and the distance should be unlimited (because the objects distance is not limited).
- I initially thought that if no positionOffet was specified in the xml there could still be a positionOffset set. I can see that's not the case now but it might be good to add a test for this?
- I can see why you've done it this way, but it's disappointing that it's not possible to share more with the regular position class.

There's a few formatting errors, which should be easy to fix; I've got a few other comments that I'll leave in-line.

File: src/private/xml_parser.cpp [hunk 1]
```diff
@@ -193,6 +193,13 @@ namespace adm {
       addOptionalReferences<AudioPackFormatId>(node, "audioPackFormatIDRef", audioObject, objectPackFormatRefs_, &parseAudioPackFormatId);
       addOptionalReferences<AudioTrackUidId>(node, "audioTrackUIDRef", audioObject, objectTrackUidRefs_, &parseAudioTrackUidId);
       setOptionalElement<AudioObjectInteraction>(node, "audioObjectInteraction", audioObject, &parseAudioObjectInteraction);
+
+      if(guessCartesianFlag(node, "positionOffset") == false) {
```
Comment (tomjnixon) [inline src/private/xml_parser.cpp:197]: It's a personal preference, but I would find this easier to parse if it was just (`if(guessCartesianFlag(node, "positionOffset")) {` with the cases swapped around.

File: src/private/xml_parser.cpp [hunk 2]
```diff
@@ -624,8 +631,8 @@ namespace adm {
       return frequency;
     }
 
-    Cartesian guessCartesianFlag(NodePtr node) {
-      auto element = detail::findElement(node, "position");
+    Cartesian guessCartesianFlag(NodePtr node, std::string elementName) {
```
Comment (tomjnixon) [inline src/private/xml_parser.cpp:634]: Using `const std::string &elementName` would copying a string.

File: src/elements/position_offset.cpp [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    y_ = boost::none;
+  }
+  void CartesianPositionOffset::unset(detail::ParameterTraits<ZOffset>::tag) {
+    z_ = boost::none;
+  }
+
+  // ---- FREE FUNCTIONS ---- //
+
+  bool isSpherical(const PositionOffset& offset) {
+    if (offset.which() == 0) {
```
Comment (tomjnixon) [inline src/elements/position_offset.cpp:154]: I know this is the same as the in `Position`, but I'd rather `.which()` was not used to test variant types, as it depends on the order. `return offset.type() == typeid(SphericalPositionOffset);` would be my personal preference (same below).

**Comment by kristianhentschelbbc** [comment]
Thanks for the review!

>     * I'm not sure about the defaults and limits of the offsets; the specification doesn't give either. Defaulting to zero makes sense. If we're going to keep the limits then I'd say the elevation offset should be -180 to 180 (difference between sources at -90 and 90), and the distance should be unlimited (because the objects distance is not limited).

Have updated these to -180, 180 for elevation and -360, 360 for azimuth to cover the range of the corresponding Position attributes.

>     * I initially thought that if no positionOffet was specified in the xml there could still be a positionOffset set. I can see that's not the case now but it might be good to add a test for this?

Added the check to the basic object parsing test now so at least it's 'documented' that way. (Checking with Matt as I don't quite remember why we decided to make it optional, not defaulted)

>     * I can see why you've done it this way, but it's disappointing that it's not possible to share more with the regular position class.

Yes, Matt and I discussed this a bit; hence my frustration with the vast amount of new (largely copy-pasted) code. This will be a good target for the new auto base class, especially if the template wrappers can also be abstracted, but I think that's maybe a separate feature.

> There's a few formatting errors, which should be easy to fix

I've run clang-format again and it seems happy now with one indentation change; did you spot any specific errors; or do you mean some of the XML files using tabs, and some using spaces (I think the ones created by the library have tabs...)?

**Comment by tomjnixon** [comment]
> >
> > * I'm not sure about the defaults and limits of the offsets; the specification doesn't give either. Defaulting to zero makes sense. If we're going to keep the limits then I'd say the elevation offset should be -180 to 180 (difference between sources at -90 and 90), and the distance should be unlimited (because the objects distance is not limited).
> >
> 
> Have updated these to -180, 180 for elevation and -360, 360 for azimuth to cover the range of the corresponding Position attributes.

Thanks, sounds sensible.

> >
> > * I initially thought that if no positionOffet was specified in the xml there could still be a positionOffset set. I can see that's not the case now but it might be good to add a test for this?
> >
> 
> Added the check to the basic object parsing test now so at least it's 'documented' that way. (Checking with Matt as I don't quite remember why we decided to make it optional, not defaulted)

Great, I think optional makes sense, as it makes it easier to check that a file doesn't use this feature.

> > There's a few formatting errors, which should be easy to fix
> 
> I've run clang-format again and it seems happy now with one indentation change; did you spot any specific errors; or do you mean some of the XML files using tabs, and some using spaces (I think the ones created by the library have tabs...)?

Sorry about this, i was thrown off by the missing space after 'if' here https://github.com/ebu/libadm/pull/98/files#diff-bee56ca68417e7ed0531c182bf9520ba80f4a02df26a9192189e8f3a8c4fecd2R198 , which is inside a `clang-format off` block, and the missing newline at the end of `position_offset_tests.cpp`, which apparently isn't fixed by clang-format. I don't think those need fixing, that's just why I thought it might need doing.

**Comment by tomjnixon** [comment]
Merged through #120 and #125

**Pull request closed without merging**